awx/main/tasks/receptor.py

```diff
@@ -266,139 +267,119 @@ def __init__(self, task, runner_params=None):
             self.runner_params['only_transmit_kwargs'] = True
 
     def run(self):
-        # We establish a connection to the Receptor socket
-        receptor_ctl = get_receptor_ctl()
-
         res = None
+        unit_id = None
         try:
-            res = self._run_internal(receptor_ctl)
+            res, unit_id = self._run()
+
+            if res.status == 'error':
+                self.handle_error(unit_id)
+
             return res
         finally:
             # Make sure to always release the work unit if we established it
-            if self.unit_id is not None and settings.RECEPTOR_RELEASE_WORK:
+            if unit_id is not None and settings.RECEPTOR_RELEASE_WORK:
                 try:
-                    receptor_ctl.simple_command(f"work release {self.unit_id}")
+                    self.receptor_ctl.simple_command(f"work release {unit_id}")
                 except Exception:
-                    logger.exception(f"Error releasing work unit {self.unit_id}.")
+                    logger.exception(f"Error releasing work unit {unit_id}.")
+
+    def _run(self):
+        payload_writer, payload_reader = socket.socketpair()
+
+        # This ThreadPoolExecutor runs for the duration of the job.
+        # The cancel_func pattern is intended to guard againsnt any situation where we may
+        # end up stuck while reading or writing from the receptor socket.
+        with concurrent.futures.ThreadPoolExecutor(max_workers=3) as executor:
+            # Launch transmitter and submitter at the same time so the payload
+            # can be read by receptor while it's still being written.
+            transmitter_future = executor.submit(self.transmitter, payload_writer)
+
+            # a file-like interface for receptor to read the payload from
+            payload_reader_file = payload_reader.makefile('rb')
+            submitter_future = executor.submit(self.submitter, payload_reader_file)
+
+            def cancel_func():
+                payload_writer.shutdown(socket.SHUT_RDWR)
+                payload_writer.close()
+                payload_reader.shutdown(socket.SHUT_RDWR)
+                payload_reader.close()
+                payload_reader_file.close()
+                self.receptor_ctl._socket.shutdown(socket.SHUT_RDWR)
+                self.receptor_ctl._socket.close()
+                self.receptor_ctl._sockfile.close()
```

Comment:
Somewhat of a tangent - I have many thought it would be better to call `self.receptor_ctl.close()`, to replace the 3 lines above here. It's written defensively enough, and would be more clear / DRY.

Reply:
`self.receptor_ctl._socket.shutdown(socket.SHUT_RDWR)` is not call in receptorctl's close, so we could keep that line. But the other 2 lines could be replaced